Log Uncaught Errors & Unhandled Promise Rejections In Main Process

gasi opened the pull request:
- [x] Add `electron-unhandled` dependency:
  - ~~Ensures errors are normalized~~ (disabled to prevent serializing non-errors that are thrown and leaking information)
  - Distinguishes between main and renderer processes
  - Allows suppression of error dialog
- [x] Log uncaught errors and unhandled promise rejections in main process
- [x] Tested using unguarded `throw new TyperError(…)` and `Promise.reject(…)` in `setTimeout` after `app` `ready` event.
- [x] Extract `Privacy` module that centralizes how we redact sensitive information such as phone numbers, group IDs, and user file paths.
- [x] Add `eslint-plugin-mocha` to disallow exclusive tests using `*.only`.

Fixes #2019.

scottnonnenberg commented on main.js:
```diff
@@ -1,3 +1,4 @@
+const addUnhandledErrorHandler = require('electron-unhandled');
```
I would like us to separate built-ins (`fs`, `path`) from third-party node modules, from our own stuff. Can we move this to the node modules location?

gasi replied:
Sure, no problem.

scottnonnenberg commented on main.js:
```diff
@@ -22,6 +23,16 @@ const logging = require('./app/logging');
+  logger: (error) => {
+    console.log(
+      'main.js: Uncaught error or unhandled promise rejection:',
+      error && error.stack ? error.stack : error
```
This might be a good opportunity to spend the time to remove the current directory from any paths on the call stack, so sensitive information doesn't get into the logs.

gasi replied:
Good idea. Let me look into that.

scottnonnenberg commented on main.js:
```diff
@@ -415,6 +426,7 @@ app.on('ready', () => {
   logging.initialize().catch((error) => {
     loggingSetupError = error;
   }).then(() => {
+  /* eslint-enable more/no-then */
```
What's happening here? Despite the title of the commit ('reducing the scope'), it seems like we're increasing its scope since we're enabling it higher in the file.

gasi replied:
Yes, we reduce the places where we can use `then`, encouraging new code to use `async` / `await`.

gasi commented:
@scottnonnenberg @scottnonnenberg-signal: Thanks for the code review. Addressed your feedback, incl. redacting sensitive file paths. Didn’t rebase but here’s a link to latest changes just in case: https://github.com/signalapp/Signal-Desktop/pull/2110/files/e3c9a81ba006f7fbebe9fe2e14299423effac665..75208f89d075de69dbfa3f08cf2ccd81e33b926d

scottnonnenberg commented on js/background.js:
```diff
@@ -416,7 +416,7 @@
         var error = c.validateNumber();
         if (error) {
-            console.log('Invalid contact received', error && error.stack ? error.stack : error);
+            console.log('Invalid contact received:', Errors.toLogFormat(error));
```
I just had a look at the `ensure-error` code, and it doesn't have the same behavior as our previous approach. Before, we looked for `.stack` and if we didn't find it, we would rely on the `toString()` of whatever we got. `ensure-error` does a `util.inspect()` on the incoming thing and creates a new callstack.

Seems like in most situations we'd be alright, perhaps even better than we were before. At least in terms of debuggability. But in some cases we'd show a whole lot more data than we did before since `util.inspect()` will enumerate keys while `toString()` didn't. That might have privacy implications.

gasi replied:
Thanks for digging in. I saw the behavior slightly changed but didn’t notice any exposure so far. Let’s keep an eye on it during the beta.

scottnonnenberg commented on js/modules/types/errors.js:
```diff
@@ -0,0 +1,9 @@
+//      toLogFormat :: Error -> String
+exports.toLogFormat = (error) => {
+  const normalizedError = ensureError(error);
+  return Privacy.redactAll(normalizedError.stack);
```
Do we need to redact here, if we're already redacting in `logging.js`?

gasi replied:
Sure; I must have added this before the general redaction in the logging module.

scottnonnenberg commented on js/modules/privacy.js:
```diff
@@ -0,0 +1,53 @@
+const GROUP_ID_PATTERN = /(group\()([^)]+)(\))/g;
+
+const APP_ROOT_PATH = Path.join(__dirname, '..', '..', '..');
+const APP_ROOT_PATH_PATTERN = new RegExp(APP_ROOT_PATH, 'g');
```
We're creating a regular expression from an arbitrary incoming path. Seems like we may want to put some protections in here to prevent crashes. Or perhaps rely on a recursive string replace instead of using regular expressions.

Also, doesn't the `path.join()` code above create a path ending with `../../..`? `path.resolve()` should help with that.

gasi replied:
Re: `RegExp`. Good call on being more defensive around creating the `RegExp` from a user supplied file path. Followed JDD’s advice and added `_.escapeRegExp`: https://github.com/lodash/lodash/issues/1084#issuecomment-86698786. Wish there was simply a `String::replaceAll` function.

Re: `path.join`. Confirmed it does the right thing and normalizes the path to not have `../../..`. Otherwise, the tests wouldn’t have passed anyways. I prefer `path.join` to `path.resolve` if it meets our needs as it is a pure function. `path.resolve` falls back `process.cwd()`.

gasi replied:
FWIW, PureScript offers a straightforward `String::replaceAll` function: https://pursuit.purescript.org/packages/purescript-strings/3.5.0/docs/Data.String#v:replaceAll

The implementation shows that it’s non-trivial: https://github.com/purescript/purescript-strings/blob/06e048c1a6026e3fd6dd9aff0a2d26da23cec783/src/Data/String.js#L114-L120

😜

gasi commented:
Thanks for pointing some edge cases I didn’t handle. Submitted changes to address these. I’d appreciate another look — thanks! https://github.com/signalapp/Signal-Desktop/pull/2110/files/75208f89d075de69dbfa3f08cf2ccd81e33b926d..2d02d620f0b086fce99d117e1eab24516b1a407a

gasi commented:
@scottnonnenberg: Ready for final review with `ensure-error` removed: https://github.com/signalapp/Signal-Desktop/pull/2110/files/2d02d620f0b086fce99d117e1eab24516b1a407a..d97fcecea2995e46c61aa8e96626c01d85f72961

gasi commented:
Rebased and pushed. Changes: Rebasing to fix `.only` leftover (thanks!) and adding `mocha` ESLint plugin 😄

gasi commented:
Rebased against after merging #2111 and #2112.

gasi commented:
Added one more commit to update our `electron-unhandled` fork and not even use `ensure-error` and `clean-error`. Will merge now.
